pbutil/decode.go

```diff
@@ -15,15 +15,38 @@
 package pbutil
 
 import (
-	"encoding/binary"
-	"errors"
 	"io"
 
+	"google.golang.org/protobuf/encoding/protodelim"
 	"google.golang.org/protobuf/proto"
 )
 
-// TODO: Give error package name prefix in next minor release.
-var errInvalidVarint = errors.New("invalid varint32 encountered")
+type countingReader struct {
+	underlying io.Reader
+	read       int
+}
+
+func (c *countingReader) Read(p []byte) (n int, err error) {
+	n, err = c.underlying.Read(p)
+	if n > 0 {
+		c.read += n
+	}
+	return n, err
+}
+
+func (c *countingReader) ReadByte() (byte, error) {
```

Comment:
Small question: Why have a singular ReadByte method that's not used by anything?  Does something in the bowels of the Protocol Buffer implementation do a runtime type switch to check if the `io.Reader` also implements this method?

If this is used by Protocol Buffer internals, could you add a comment on this method to explain that and where and how it's used?

Reply:
The ReadByte method is not unused. If you remove ReadByte, the code will fail to compile:

./decode.go:67:27: cannot use cr (variable of type *countingReader) as protodelim.Reader value in argument to opts.UnmarshalFrom: *countingReader does not implement protodelim.Reader (missing method ReadByte)

I have added a comment to indicate which interface these methods implement.

Reply:
The pitfalls of having two `Reader`s.  I casually assumed the other one was referring to `io.Reader` and not a local interface types.  Thanks for clarifying1